nucleus/flashlight/framework/src/main/java/org/glassfish/flashlight/impl/client/ProbeProviderClassFileTransformer.java

```diff
@@ -79,141 +78,94 @@
 
     ///////////////  instance variables  //////////////////
     // Don't hold a strong ref to the class, it will prevent entries in the weak map from being reclaimed
-    private final WeakReference<Class> providerClassRef;
-    private String providerClassName = null;
-    private Map<String, FlashlightProbe> probes = new HashMap<String, FlashlightProbe>();
-    private ClassWriter cw;
-    private volatile boolean enabled = false;
-    private boolean allProbesTransformed = true;
+    private final WeakReference<Class<?>> providerClassRef;
+    private final String providerClassName;
+    private final Map<String, FlashlightProbe> probes = new ConcurrentHashMap<>();
+    private AtomicBoolean flaggedForUpdate = new AtomicBoolean(false);
     private boolean transformerAdded = false;
     private int count = 0;  // Only used for debug so we can look at the before/after class dumps for each iteration
     ///////////////  static variables  //////////////////
-    // weak map here so the classes don't prevent app classloaders from being reclaimed
-    private static Map<Class, ProbeProviderClassFileTransformer> instances =
-        new WeakHashMap<Class, ProbeProviderClassFileTransformer>();
+    // uses String as key so no reference to Class is held => allow being reclaimed
+    private static Map<String, ProbeProviderClassFileTransformer> instances = new ConcurrentHashMap<>();
```

Comment:
Suspecting memory leak, @jbee may you share more detail.

Reply:
I think you are technically right. The `ProbeProviderClassFileTransformer` instance and the `String` for the key are never cleaned even if the `Class` they are fore got reclaimed. I did not address it since there are so few classes actually used and the chance of reclaiming is quite superficial but then this as far as I know is a feature accessible to applications so theoretically there could be many such classes. I am going to add a cleanup to the worker thread loop.